Action/Server refactors

Profpatsch opened the pull request:
A bunch of refactors that fell out of a study session of the module list generation code.

The original goal (at Munihac) was to change the sort order, so that the home module of a target is listed first, but we didn’t get that far.

So this is a bunch of refactors that do not (or should not) change any behavior. Any changes will be submitted in separate PRs.

Profpatsch committed:
Action/Server: Document grab* functions

Profpatsch committed:
Action/Server: Pass url options as one argument

The `local` and `haddock` booleans are only used for determining the
URLs to generate, so let’s make that clear.

Profpatsch committed:
Action/Server: enable -Wall & fix warnings

Profpatsch committed:
Action/Server/showFroms: document & upname a little

Profpatsch committed:
Action/Server/showResults: annotate functions in where

Profpatsch committed:
Action/Server/dedupeTake: rename to takeAndGroup & document

The function does not really deduplicate the elements, it takes and
groups. :)

Profpatsch commented:
cc @smatting

ndmitchell commented on src/Action/Server.hs:
```diff
@@ -1,7 +1,10 @@
 {-# LANGUAGE ViewPatterns, TupleSections, RecordWildCards, ScopedTypeVariables, PatternGuards #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# OPTIONS_GHC -Wall #-}
```
I generally find this a bad thing to put in the source code directly - what is in Wall changes, and not all the things in there are a good idea.

Profpatsch replied:
I didn’t want to fix warnings in the whole codebase for now, so I only scoped it to this module; do you want to just remove `-Wall` or have something else in mind?

Profpatsch replied:
I opted for manuall disabling the shadowing warnings & pattern warnings

ndmitchell commented on src/Action/Server.hs:
```diff
@@ -30,7 +33,7 @@ import System.Time.Extra
 import Data.Time.Calendar
 import System.IO.Unsafe
-import Numeric.Extra
+import Numeric.Extra hiding (log)
```
These look like they are only heard to satisfy shadow warnings, which seems unnnecessary, as shadowing is not a bad thing.

Profpatsch replied:
I find the scoping warning very helpful to notice accidental bugs, but ofc it’s a bit of menial work.

ndmitchell commented on src/Action/Server.hs:
```diff
@@ -1,7 +1,10 @@
+{-# LANGUAGE NamedFieldPuns #-}
 
-module Action.Server(actionServer, actionReplay, action_server_test_, action_server_test) where
+module Action.Server(actionServer, actionReplay, action_server_test_, action_server_test, takeAndGroup) where
```
Why export takeAndGroup? It doesn't look like it's used?

Profpatsch replied:
Ah, yeah, that was just to try it out in the repl

ndmitchell commented on src/Action/Server.hs:
```diff
@@ -60,11 +63,12 @@ actionServer cmd@Server{..} = do
+    _ <- evaluate spawned
     dataDir <- maybe getDataDir pure datadir
-    haddock <- maybe (pure Nothing) (fmap Just . canonicalizePath) haddock
+    haddock' <- maybe (pure Nothing) (fmap Just . canonicalizePath) haddock
```
I actually thing this is more dangerous than before. Now we have satisfied this warning if you accidentally type `haddock` on the next line it works and goes wrong. When we shadowed the variable, we eliminated that problem by construction.

ndmitchell commented on src/Action/Server.hs:
```diff
@@ -79,6 +83,7 @@ actionReplay Replay{..} = withBuffering stdout NoBuffering $ do
             evaluate $ rnf res
             putChar '.'
     putStrLn $ "\nTook " ++ showDuration t ++ " (" ++ showDuration (t / intToDouble (repeat_ * length qs)) ++ ")"
+actionReplay _ = error "should not happen"
```
Before, if we got here it would print an error message including the line number and everything. Now, we're relying on stack traces that are much more fickle.

ndmitchell commented on src/Action/Server.hs:
```diff
@@ -161,72 +178,86 @@ replyServer log local links haddock store cdn home htmlDir scope Input{..} = cas
+            | otherwise = f (k:keys) (Map.insert k [x] mp) xs
             where k = key x
 
+data UrlOpts = IsHaddockUrl | IsLocalUrl | IsOtherUrl
```
I think the normal way you'd expect this would be without the `Is` prefix

ndmitchell commented on src/Action/Server.hs:
```diff
@@ -161,72 +178,86 @@ replyServer log local links haddock store cdn home htmlDir scope Input{..} = cas
+            | otherwise = f (k:keys) (Map.insert k [x] mp) xs
             where k = key x
 
+data UrlOpts = IsHaddockUrl | IsLocalUrl | IsOtherUrl
```
Would be great if this data type had some comments as to what each of the types of url looked like or represented (I appreciate this was missing before, but now we have the type, it should be easy to add)

Profpatsch replied:
I don’t actually know what these mean semantically, I just refactored. I’ll leave it up to you to document the cases :)

ndmitchell reviewed (commented):
Thanks very much for the PR! I suggest I review/land this one first, before looking at the others, so we can get through them in order. Let me know if that doesn't seem reasonable.

A few notes - I think the use of Wall makes the code worse in a few ways, but better in a bunch, so I think just winding back a few of them makes sense. Overall, looks good and some nice improvements. Thanks!

Profpatsch committed:
Action/Server: Disable unwanted warnings

Upstream does not think shadowing should be avoided, so let’s undo the
changes to shadowing.
Same with incomplete pattern warnings, we’ll just let it crash for
now.

Also drop the `Is*` prefix for the URL constructors.

ndmitchell commented:
I seem to be getting failures on newer GHC:

```
src/Action/Server.hs:217:21: error: [-Wincomplete-uni-patterns, -Werror=incomplete-uni-patterns]
Error:     Pattern match(es) are non-exhaustive
    In a lambda abstraction:
        Patterns of type ‘[Target]’ not matched: []
    |
217 |     forM_ results $ \is@(Target{..}:_) -> do
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^...
```

Happy with removing the wall, or fixing it. Also wouldn't be averse to using NonEmpty in the right places to fix it too.

Profpatsch committed:
Action/Server: remove -Wall again, to prevent -Werror killing CI

CI unfortunately uses -Werror and tests against more modern versions
of GHC, so any new errors will only appear on CI.

Profpatsch commented:
I opted for removing `-Wall` for now since I think it would interfere with the other PR’s commits too much and lead to a big rebase.

Profpatsch commented:
ok looks like the others are unrelated

on MacOS:

```
FAILED ON MAC
{"message":"API rate limit exceeded for 199.7.166.17. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}
```

on Ubuntu:

```
cabal: Failed to build wai-logger-2.4.0. The failure occurred during the
configure step. The exception was:
dieVerbatim: user error (cabal: '/home/runner/.ghcup/bin/ghc' exited with an
error:
/home/runner/.cabal/store/ghc-9.0.2/cabal-doctest-1.0.9-c5067c39fbeb5bff36b6ed7706ee31a87e1421142e9fe6d94eb8209fc65f5188/lib/libHScabal-doctest-1.0.9-c5067c39fbeb5bff36b6ed7706ee31a87e1421142e9fe6d94eb8209fc65f5188.a(Doctest.o)(.text+0x3066):
error: undefined reference to
'base_GHCziShow_zdfShowZLz2cUZRzuzdcshowList1_closure'
…
```

Profpatsch commented:
Okay, I might have some time to work on these again, I rebased. I think most of the comments have been fixed, PTAL.

Profpatsch commented:
Now all CI runs fail because of the copyright year :(

Bodigrim commented:
@Profpatsch could you please rebase, so that we can give CI another go? It should be in working order now.
